echopype/echodata/echodata.py

```diff
@@ -315,114 +314,167 @@ def update_platform(
+            # fmt: on
+            max_index = min(
+                np.searchsorted(
+                    sorted_external_time,
+                    self["Sonar/Beam_group1"]["ping_time"].max(),
+                    side="right",
+                ),
+                len(sorted_external_time) - 1,
+            )
+            return external_ds.sel(
+                {
+                    ext_time_dim_name: np.logical_and(
+                        sorted_external_time[min_index] <= external_ds[ext_time_dim_name],
+                        external_ds[ext_time_dim_name] <= sorted_external_time[max_index],
+                    )
+                }
+            )
 
-        platform = self["Platform"]
-        platform = platform.drop_dims(["time1"], errors="ignore")
-        # drop_dims is also dropping latitude, longitude and sentence_type why?
-        platform = platform.assign_coords(time1=extra_platform_data[time_dim].values)
+        # History attribute to be included in each updated variable
         history_attr = f"{datetime.datetime.utcnow()} +00:00. Added from external platform data"
         if extra_platform_data_file_name:
             history_attr += ", from file " + extra_platform_data_file_name
-        time1_attrs = {
-            **self._varattrs["platform_coord_default"]["time1"],
-            **{"history": history_attr},
-        }
-        platform["time1"] = platform["time1"].assign_attrs(**time1_attrs)
-
-        platform_vars_sourcenames = {
-            "pitch": ["pitch", "PITCH"],
-            "roll": ["roll", "ROLL"],
-            "vertical_offset": ["heave", "HEAVE", "vertical_offset", "VERTICAL_OFFSET"],
-            "latitude": ["lat", "latitude", "LATITUDE"],
-            "longitude": ["lon", "longitude", "LONGITUDE"],
-            "water_level": ["water_level", "WATER_LEVEL"],
-        }
 
-        dropped_vars_target = platform_vars_sourcenames.keys()
-        dropped_vars = []
-        for var in dropped_vars_target:
-            if var in platform and (~platform[var].isnull()).all():
-                dropped_vars.append(var)
-        if len(dropped_vars) > 0:
+        platform = self["Platform"]
+
+        # Drop variable_mappings items where either the Platform group or
+        # extra_platform_data don't contain the corresponding variables or contain only nan data
+        mappings_expanded = {}
+        for platform_var, external_var in variable_mappings.items():
+            if platform_var in platform:
+                platform_validvalues = not platform[platform_var].isnull().all()
+                ext_present, ext_validvalues, ext_time_dim_name = _extvar_properties(
+                    extra_platform_data, external_var
+                )
+                if ext_present and ext_validvalues:
+                    mappings_expanded[platform_var] = (
+                        external_var,
+                        ext_time_dim_name,
+                        platform_validvalues,
+                    )
```

Comment:
I found myself flipping back multiple times to this block to check what index 0, 1, 2 means. How about changing this to a dictionary of dictionaries? Then the code below can explicitly use "external_var", "ext_time_dim_name", and "platform_validvalues" in the places where you have `v[0]`, `v[1]`, `v[-1]`.

Reply:
Yeah ... I don't disagree. I ran out of steam to replace it with a dict. Let me see if I can do it now.

Reply:
Ok, I've pushed these changes just now. The code is much more readable now!